File: packages/gg-sandbox-java/src/gg/internal/Read.java
package gg.internal;

import gg.board.Board;
import gg.context.Context;
import gg.delegation.Delegation;
import gg.docs.Docs;
import gg.files.Files;
import gg.memories.Memories;
import gg.programs.Programs;
import gg.shell.Shell;
import gg.tasks.Tasks;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.OptionalInt;

/**
 * <b>Reading the wire</b> — every {@link Value} gg answers with, as the Java value this SDK's
 * signatures promise.
 *
 * <p>Nothing here is model-facing, and nothing here is clever: each function is the one place that
 * knows a field's name on the wire, so a rename is one edit rather than a search. A field name is
 * the <b>WIT</b> name verbatim, hyphens and all, because that is what
 * {@code crates/gg/src/sandbox/membrane/wire.*.rs} writes and gg owns both ends of it.
 *
 * <p>The three shapes worth naming are the ones a language with no {@code undefined} has to decide
 * about — a field the wire may leave out becomes {@link Optional} or {@link OptionalInt}, a
 * discriminated union becomes a real Java subtype, and a fixed choice becomes an enum constant
 * rather than the case name it arrived as.
 */
public final class Read {
    private Read() {
    }

    // -------------------------------------------------------------------------------------------
    // The building blocks
    // -------------------------------------------------------------------------------------------

    /** A whole number the wire may have left out. */
    public static OptionalInt optionalInt(Value value) {
        return value.absent() ? OptionalInt.empty() : OptionalInt.of(value.integer());
    }

    /** Text the wire may have left out. */
    public static Optional<String> optionalText(Value value) {
        return value.absent() ? Optional.empty() : Optional.of(value.text());
    }

    /** A list of text, as an unmodifiable list. */
    public static List<String> texts(Value value) {
        List<String> out = new ArrayList<>(value.size());
        for (int index = 0; index < value.size(); index++) {
            out.add(value.at(index).text());
        }
        return List.copyOf(out);
    }

    // -------------------------------------------------------------------------------------------
    // The results
    // -------------------------------------------------------------------------------------------

    /** What a command reported. */
    public static Shell.ShellOutput shellOutput(Value value) {
        return new Shell.ShellOutput(optionalInt(value.get("exit-code")),
                value.get("output").text(), value.get("truncated").flag());
    }

    /** A read, narrowed to the arm gg tagged it with. */
    public static Files.FileRead fileRead(Value value) {
        Value read = value.get("value");
        if ("image".equals(value.get("case").text())) {
            return new Files.ImageFile(read.get("media-type").text(), read.get("label").text(),
                    read.get("bytes").integer(), read.get("shown").flag(),
                    optionalText(read.get("not-shown-reason")));
        }
        return new Files.TextFile(read.get("contents").text(), read.get("first-line").integer(),
                read.get("last-line").integer(), read.get("total-lines").integer(),
                read.get("byte-truncated").flag());
    }

    /** One directory entry. */
    public static Files.DirEntry dirEntry(Value value) {
        return new Files.DirEntry(value.get("name").text(), entryKind(value.get("kind").text()));
    }

    /** Every directory entry in a list. */
    public static List<Files.DirEntry> dirEntries(Value value) {
        List<Files.DirEntry> out = new ArrayList<>(value.size());
        for (int index = 0; index < value.size(); index++) {
            out.add(dirEntry(value.at(index)));
        }
        return List.copyOf(out);
    }

    /** Every line a search matched. */
    public static List<Files.SearchMatch> searchMatches(Value value) {
        List<Files.SearchMatch> out = new ArrayList<>(value.size());
        for (int index = 0; index < value.size(); index++) {
            Value match = value.at(index);
            out.add(new Files.SearchMatch(match.get("path").text(), match.get("line").integer(),
                    match.get("text").text()));
        }
        return List.copyOf(out);
    }

    /** The memory budget. */
    public static Memories.MemoryUsage memoryUsage(Value value) {
        return new Memories.MemoryUsage(value.get("count").integer(),
                optionalInt(value.get("max-count")), value.get("total-chars").integer(),
                optionalInt(value.get("max-total-chars")), optionalInt(value.get("index-chars")),
                optionalInt(value.get("max-index-chars")));
    }

    /** Every memory a search matched. */
    public static List<Memories.MemoryHit> memoryHits(Value value) {
        List<Memories.MemoryHit> out = new ArrayList<>(value.size());
        for (int index = 0; index < value.size(); index++) {
            Value hit = value.at(index);
            out.add(new Memories.MemoryHit(hit.get("name").text(), hit.get("description").text(),
                    hit.get("matched").integer(), hit.get("occurrences").integer(),
                    hit.get("excerpt").text()));
        }
        return List.copyOf(out);
    }

    /** One page of a documentation search. */
    public static Docs.DocSearch docSearch(Value value) {
        Value found = value.get("hits");
        List<Docs.DocHit> hits = new ArrayList<>(found.size());
        for (int index = 0; index < found.size(); index++) {
            Value hit = found.at(index);
            hits.add(new Docs.DocHit(hit.get("key").text(), docKind(hit.get("kind").text()),
                    hit.get("module").text(), hit.get("name").text(),
                    hit.get("summary").text()));
        }
        return new Docs.DocSearch(value.get("total").integer(), value.get("offset").integer(),
                List.copyOf(hits));
    }

    /** The task budget. */
    public static Tasks.TaskUsage taskUsage(Value value) {
        return new Tasks.TaskUsage(value.get("count").integer(),
                value.get("max-tasks").integer());
    }

    /** The board budget. */
    public static Board.BoardUsage boardUsage(Value value) {
        return new Board.BoardUsage(value.get("epics").integer(),
                value.get("max-epics").integer(), value.get("issues").integer(),
                value.get("max-issues").integer());
    }

    /** An epic that was just created. */
    public static Board.EpicCreated epicCreated(Value value) {
        return new Board.EpicCreated(value.get("id").text(), boardUsage(value.get("board")));
    }

    /** An issue that was just created. */
    public static Board.IssueCreated issueCreated(Value value) {
        return new Board.IssueCreated(value.get("id").text(), boardUsage(value.get("board")));
    }

    /** What a reclaim freed. */
    public static Context.ReclaimReport reclaimReport(Value value) {
        return new Context.ReclaimReport(value.get("items").integer(),
                value.get("reclaimed-tokens").integer(), texts(value.get("paths")),
                value.get("detail").text());
    }

    /** What an archive search found. */
    public static Context.ArchiveSearch archiveSearch(Value value) {
        Value found = value.get("hits");
        List<Context.ArchiveHit> hits = new ArrayList<>(found.size());
        for (int index = 0; index < found.size(); index++) {
            Value hit = found.at(index);
            hits.add(new Context.ArchiveHit(hit.get("seq").integer(),
                    messageRole(hit.get("role").text()), hit.get("text").text()));
        }
        return new Context.ArchiveSearch(value.get("archive-empty").flag(), List.copyOf(hits));
    }

    /** A child agent's handle. */
    public static Delegation.SubagentHandle subagentHandle(Value value) {
        return new Delegation.SubagentHandle(value.get("id").text(), value.get("slot").text(),
                value.get("model-id").text());
    }

    /** Every child agent's result. */
    public static List<Delegation.SubagentResult> subagentResults(Value value) {
        List<Delegation.SubagentResult> out = new ArrayList<>(value.size());
        for (int index = 0; index < value.size(); index++) {
            Value result = value.at(index);
            out.add(new Delegation.SubagentResult(result.get("id").text(),
                    optionalText(result.get("status")).map(Read::agentEnding),
                    result.get("summary").text()));
        }
        return List.copyOf(out);
    }

    /** Every program this session has run. */
    public static List<Programs.ProgramSummary> programSummaries(Value value) {
        List<Programs.ProgramSummary> out = new ArrayList<>(value.size());
        for (int index = 0; index < value.size(); index++) {
            Value program = value.at(index);
            out.add(new Programs.ProgramSummary(program.get("id").text(),
                    program.get("turn").integer(),
                    program.get("lines").integer(), program.get("chars").integer(),
                    program.get("ok").flag(), optionalText(program.get("error"))));
        }
        return List.copyOf(out);
    }

    // -------------------------------------------------------------------------------------------
    // The fixed choices
    // -------------------------------------------------------------------------------------------

    /** What a directory entry is, from the case name the wire used. */
    private static Files.EntryKind entryKind(String wire) {
        return switch (wire) {
            case "file" -> Files.EntryKind.FILE;
            case "directory" -> Files.EntryKind.DIRECTORY;
            default -> Files.EntryKind.OTHER;
        };
    }

    /** Who said an archived message, from the case name the wire used. */
    private static Context.MessageRole messageRole(String wire) {
        return switch (wire) {
            case "system" -> Context.MessageRole.SYSTEM;
            case "assistant" -> Context.MessageRole.ASSISTANT;
            case "tool" -> Context.MessageRole.TOOL;
            default -> Context.MessageRole.USER;
        };
    }

    /** Which kind a documentation entry is, from the word the wire used. */
    private static Docs.DocKind docKind(String wire) {
        return switch (wire) {
            case "module" -> Docs.DocKind.MODULE;
            case "type" -> Docs.DocKind.TYPE;
            default -> Docs.DocKind.FUNCTION;
        };
    }

    /** How a child agent ended, from the case name the wire used. */
    private static Delegation.AgentEnding agentEnding(String wire) {
        return switch (wire) {
            case "completed" -> Delegation.AgentEnding.COMPLETED;
            case "exhausted" -> Delegation.AgentEnding.EXHAUSTED;
            case "timed-out" -> Delegation.AgentEnding.TIMED_OUT;
            case "auth-error" -> Delegation.AgentEnding.AUTH_ERROR;
            case "limit-exceeded" -> Delegation.AgentEnding.LIMIT_EXCEEDED;
            default -> Delegation.AgentEnding.MODEL_ERROR;
        };
    }
}

File: packages/gg-sandbox-kotlin/src/gg/internal/Read.kt
package gg.internal

import gg.board.BoardUsage
import gg.board.EpicCreated
import gg.board.IssueCreated
import gg.context.ArchiveHit
import gg.context.ArchiveSearch
import gg.context.MessageRole
import gg.context.ReclaimReport
import gg.delegation.AgentEnding
import gg.delegation.SubagentHandle
import gg.delegation.SubagentResult
import gg.docs.DocHit
import gg.docs.DocKind
import gg.docs.DocSearch
import gg.files.DirEntry
import gg.files.EntryKind
import gg.files.FileRead
import gg.files.ImageFile
import gg.files.SearchMatch
import gg.files.TextFile
import gg.memories.MemoryHit
import gg.memories.MemoryUsage
import gg.programs.ProgramSummary
import gg.shell.ShellOutput
import gg.tasks.TaskUsage

/**
 * **Reading the wire** — every [Value] gg answers with, as the Kotlin value this SDK's signatures
 * promise.
 *
 * Nothing here is model-facing, and nothing here is clever: each function is the one place that knows
 * a field's name on the wire, so a rename is one edit rather than a search. A field name is the
 * **WIT** name verbatim, hyphens and all, because that is what
 * `crates/gg/src/sandbox/membrane/wire.*.rs` writes and gg owns both ends of it.
 *
 * Three shapes are worth naming, because each is a decision Kotlin makes differently from the arm
 * this file's [Java sibling](https://docs.testcabinet.ai/gg/languages/java/) serves — a field the wire
 * may leave out becomes a **nullable** rather than an `Optional`, a discriminated union becomes an arm
 * of a **sealed interface**, and a fixed choice becomes an **enum entry** rather than the case name it
 * arrived as.
 */
internal object Read {
    // -------------------------------------------------------------------------------------------
    // The results
    // -------------------------------------------------------------------------------------------

    /** What a command reported. */
    fun shellOutput(value: Value): ShellOutput =
        ShellOutput(
            exitCode = ggOptionalInt(value.get("exit-code")),
            output = value.get("output").text(),
            truncated = value.get("truncated").flag(),
        )

    /** A read, narrowed to the arm gg tagged it with. */
    fun fileRead(value: Value): FileRead {
        val read = value.get("value")
        return if (value.get("case").text() == "image") {
            ImageFile(
                mediaType = read.get("media-type").text(),
                label = read.get("label").text(),
                bytes = read.get("bytes").integer(),
                shown = read.get("shown").flag(),
                notShownReason = ggOptionalText(read.get("not-shown-reason")),
            )
        } else {
            TextFile(
                contents = read.get("contents").text(),
                firstLine = read.get("first-line").integer(),
                lastLine = read.get("last-line").integer(),
                totalLines = read.get("total-lines").integer(),
                byteTruncated = read.get("byte-truncated").flag(),
            )
        }
    }

    /** Every directory entry in a list. */
    fun dirEntries(value: Value): List<DirEntry> =
        ggEach(value) {
            DirEntry(name = it.get("name").text(), kind = entryKind(it.get("kind").text()))
        }

    /** Every line a search matched. */
    fun searchMatches(value: Value): List<SearchMatch> =
        ggEach(value) {
            SearchMatch(
                path = it.get("path").text(),
                line = it.get("line").integer(),
                text = it.get("text").text(),
            )
        }

    /** The memory budget. */
    fun memoryUsage(value: Value): MemoryUsage =
        MemoryUsage(
            count = value.get("count").integer(),
            maxCount = ggOptionalInt(value.get("max-count")),
            totalChars = value.get("total-chars").integer(),
            maxTotalChars = ggOptionalInt(value.get("max-total-chars")),
            indexChars = ggOptionalInt(value.get("index-chars")),
            maxIndexChars = ggOptionalInt(value.get("max-index-chars")),
        )

    /** Every memory a search matched. */
    fun memoryHits(value: Value): List<MemoryHit> =
        ggEach(value) {
            MemoryHit(
                name = it.get("name").text(),
                description = it.get("description").text(),
                matched = it.get("matched").integer(),
                occurrences = it.get("occurrences").integer(),
                excerpt = it.get("excerpt").text(),
            )
        }

    /** One page of a documentation search. */
    fun docSearch(value: Value): DocSearch =
        DocSearch(
            total = value.get("total").integer(),
            offset = value.get("offset").integer(),
            hits =
                ggEach(value.get("hits")) {
                    DocHit(
                        key = it.get("key").text(),
                        kind = docKind(it.get("kind").text()),
                        module = it.get("module").text(),
                        name = it.get("name").text(),
                        summary = it.get("summary").text(),
                    )
                },
        )

    /** The task budget. */
    fun taskUsage(value: Value): TaskUsage =
        TaskUsage(count = value.get("count").integer(), maxTasks = value.get("max-tasks").integer())

    /** The board budget. */
    fun boardUsage(value: Value): BoardUsage =
        BoardUsage(
            epics = value.get("epics").integer(),
            maxEpics = value.get("max-epics").integer(),
            issues = value.get("issues").integer(),
            maxIssues = value.get("max-issues").integer(),
        )

    /** An epic that was just created. */
    fun epicCreated(value: Value): EpicCreated =
        EpicCreated(id = value.get("id").text(), board = boardUsage(value.get("board")))

    /** An issue that was just created. */
    fun issueCreated(value: Value): IssueCreated =
        IssueCreated(id = value.get("id").text(), board = boardUsage(value.get("board")))

    /** What a reclaim freed. */
    fun reclaimReport(value: Value): ReclaimReport =
        ReclaimReport(
            items = value.get("items").integer(),
            reclaimedTokens = value.get("reclaimed-tokens").integer(),
            paths = ggTextList(value.get("paths")),
            detail = value.get("detail").text(),
        )

    /** What an archive search found. */
    fun archiveSearch(value: Value): ArchiveSearch =
        ArchiveSearch(
            archiveEmpty = value.get("archive-empty").flag(),
            hits =
                ggEach(value.get("hits")) {
                    ArchiveHit(
                        seq = it.get("seq").integer(),
                        role = messageRole(it.get("role").text()),
                        text = it.get("text").text(),
                    )
                },
        )

    /** A child agent's handle. */
    fun subagentHandle(value: Value): SubagentHandle =
        SubagentHandle(
            id = value.get("id").text(),
            slot = value.get("slot").text(),
            modelId = value.get("model-id").text(),
        )

    /** Every child agent's result. */
    fun subagentResults(value: Value): List<SubagentResult> =
        ggEach(value) {
            SubagentResult(
                id = it.get("id").text(),
                status = ggOptionalText(it.get("status"))?.let(::agentEnding),
                summary = it.get("summary").text(),
            )
        }

    /** Every program this session has run. */
    fun programSummaries(value: Value): List<ProgramSummary> =
        ggEach(value) {
            ProgramSummary(
                turn = it.get("turn").integer(),
                lines = it.get("lines").integer(),
                chars = it.get("chars").integer(),
                ok = it.get("ok").flag(),
                error = ggOptionalText(it.get("error")),
            )
        }

    // -------------------------------------------------------------------------------------------
    // The fixed choices
    // -------------------------------------------------------------------------------------------

    /** What a directory entry is, from the case name the wire used. */
    private fun entryKind(wire: String): EntryKind =
        when (wire) {
            "file" -> EntryKind.FILE
            "directory" -> EntryKind.DIRECTORY
            else -> EntryKind.OTHER
        }

    /** Who said an archived message, from the case name the wire used. */
    private fun messageRole(wire: String): MessageRole =
        when (wire) {
            "system" -> MessageRole.SYSTEM
            "assistant" -> MessageRole.ASSISTANT
            "tool" -> MessageRole.TOOL
            else -> MessageRole.USER
        }

    /** Which kind a documentation entry is, from the case name the wire used. */
    private fun docKind(wire: String): DocKind =
        when (wire) {
            "module" -> DocKind.MODULE
            "type" -> DocKind.TYPE
            else -> DocKind.FUNCTION
        }

    /** Which kind a view is, from the case name the wire used. */
    /** How a child agent ended, from the case name the wire used. */
    private fun agentEnding(wire: String): AgentEnding =
        when (wire) {
            "completed" -> AgentEnding.COMPLETED
            "exhausted" -> AgentEnding.EXHAUSTED
            "timed-out" -> AgentEnding.TIMED_OUT
            "auth-error" -> AgentEnding.AUTH_ERROR
            "limit-exceeded" -> AgentEnding.LIMIT_EXCEEDED
            else -> AgentEnding.MODEL_ERROR
        }
}
